cpp/include/raft/comms/ucp_helper.hpp

```diff
@@ -0,0 +1,231 @@
+ * other fields for trace logging and cleanup.
+ */
+class ucp_request {
+ public:
+  struct ucx_context *req;
+  bool needs_release = true;
+  int other_rank = -1;
+  bool is_send_request = false;
+};
+
+// by default, match the whole tag
+static const ucp_tag_t default_tag_mask = -1;
+
+/**
+ * @brief Asynchronous send callback sets request to completed
+ */
+static void send_callback(void *request, ucs_status_t status) {
+  struct ucx_context *context = (struct ucx_context *)request;
+  context->completed = 1;
+}
+
+/**
+ * @brief Asynchronous recv callback sets request to completed
+ */
+static void recv_callback(void *request, ucs_status_t status,
+                          ucp_tag_recv_info_t *info) {
+  struct ucx_context *context = (struct ucx_context *)request;
+  context->completed = 1;
+}
+
+/**
+ * Helper class for managing `dlopen` state and
+ * interacting with ucp.
+ */
+class comms_ucp_handler {
+ public:
+  comms_ucp_handler() {
+    load_ucp_handle();
+    load_send_func();
+    load_recv_func();
+    load_free_req_func();
+    load_print_info_func();
+    load_worker_progress_func();
+  }
+
+  ~comms_ucp_handler() { dlclose(ucp_handle); }
+
+ private:
+  void *ucp_handle;
+
+  dlsym_print_info print_info_func;
+  dlsym_rec_free req_free_func;
+  dlsym_worker_progress worker_progress_func;
+  dlsym_send send_func;
+  dlsym_recv recv_func;
+
+  void load_ucp_handle() {
+    ucp_handle = dlopen("libucp.so", RTLD_LAZY | RTLD_NOLOAD | RTLD_NODELETE);
+    if (!ucp_handle) {
+      ucp_handle = dlopen("libucp.so", RTLD_LAZY | RTLD_NODELETE);
+      ASSERT(ucp_handle, "Cannot open UCX library: %s\n", dlerror());
+    }
+    // Reset any potential error
+    dlerror();
+  }
+
+  void assert_dlerror() {
+    char *error = dlerror();
+    ASSERT(error == NULL, "Error loading function symbol: %s\n", error);
+  }
+
+  void load_send_func() {
+    send_func = (dlsym_send)dlsym(ucp_handle, "ucp_tag_send_nb");
+    assert_dlerror();
+  }
+
+  void load_free_req_func() {
+    req_free_func = (dlsym_rec_free)dlsym(ucp_handle, "ucp_request_free");
+    assert_dlerror();
+  }
+
+  void load_print_info_func() {
+    print_info_func = (dlsym_print_info)dlsym(ucp_handle, "ucp_ep_print_info");
+    assert_dlerror();
+  }
+
+  void load_worker_progress_func() {
+    worker_progress_func =
+      (dlsym_worker_progress)dlsym(ucp_handle, "ucp_worker_progress");
+    assert_dlerror();
+  }
+
+  void load_recv_func() {
+    recv_func = (dlsym_recv)dlsym(ucp_handle, "ucp_tag_recv_nb");
+    assert_dlerror();
+  }
+
+  ucp_tag_t build_message_tag(int rank, int tag) const {
+    // keeping the rank in the lower bits enables debugging.
+    return ((uint32_t)tag << 31) | (uint32_t)rank;
+  }
+
+ public:
+  int ucp_progress(ucp_worker_h worker) const {
+    return (*(worker_progress_func))(worker);
+  }
+
+  /**
+   * @brief Frees any memory underlying the given ucp request object
+   */
+  void free_ucp_request(ucp_request *request) const {
```

Comment:
This will lead to memory leak if users forget to call free_ucp_request or on exceptions. Can't we use std::unique_ptr with custom deleter doing what this function does instead?

Reply:
The comms api is meant to follow closely to MPI so that we can use CUDA-aware MPI underneath does automates gtests (which be in a follow-on PR). Users are not expected to know about, nor free UCX requests on their own. The requests are freed automatically when waitall is called. This is similar to MPI.

Reply:
I see, and is this the reason we're using malloc in

https://github.com/rapidsai/raft/pull/7/files/d4aa5c5cc58ba8754fa8d61bb95132b572c223cd#diff-0835dc79c2c9697632a30a73323b374fR232

?

I don't know the mechanism to automate gtests, so if there is a clear reason we cannot avoid using malloc, please ignore my comment about malloc above (comments explaining why we should stick to old style exception unsafe malloc will be helpful).